File: tests/cross_driver/node/updateuser_smoke.js
// Cross-driver updateUser smoke — Node (mongo-node-driver).
//
// Mirrors tests/cross_driver/go/updateuser/main.go. Provisions a user
// with `orig`, rotates to `rotated` via updateUser, and asserts the
// old password no longer authenticates while the new one does.

const { MongoClient } = require('mongodb');

function fail(msg) {
  console.error(msg);
  process.exit(1);
}

function clientFor(uri, user, pwd, authSource) {
  return new MongoClient(uri, {
    auth: { username: user, password: pwd },
    authSource,
    authMechanism: 'SCRAM-SHA-256',
    serverSelectionTimeoutMS: 5000,
  });
}

async function main() {
  const uri = process.env.MONGODB_URI;
  const adminPwd = process.env.ADMIN_PASSWORD;
  if (!uri || !adminPwd) fail('MONGODB_URI and ADMIN_PASSWORD required');

  const root = clientFor(uri, 'root', adminPwd, 'admin');
  await root.connect();
  try {
    await root.db('admin').command({
      createUser: 'alice_xd',
      pwd: 'orig',
      roles: [{ role: 'read', db: 'admin' }],
    });
    await root.db('admin').command({ updateUser: 'alice_xd', pwd: 'rotated' });
  } finally {
    await root.close();
  }

  // Old password — must fail.
  let oldErr = null;
  const oldClient = clientFor(uri, 'alice_xd', 'orig', 'admin');
  try {
    await oldClient.connect();
    await oldClient.db('admin').command({ ping: 1 });
  } catch (err) {
    oldErr = err;
  } finally {
    await oldClient.close().catch(() => {});
  }
  if (!oldErr) fail('old password should not authenticate');

  // New password — must succeed.
  const newClient = clientFor(uri, 'alice_xd', 'rotated', 'admin');
  try {
    await newClient.connect();
    await newClient.db('admin').command({ ping: 1 });
  } finally {
    await newClient.close().catch(() => {});
  }

  console.log('OK');
}

main().catch((err) => fail(`uncaught: ${err && err.stack ? err.stack : err}`));

File: tests/cross_driver/java/src/main/java/com/secantus/smokes/UpdateUserSmoke.java
package com.secantus.smokes;

import com.mongodb.ConnectionString;
import com.mongodb.MongoClientSettings;
import com.mongodb.MongoCredential;
import com.mongodb.MongoSecurityException;
import com.mongodb.MongoTimeoutException;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import org.bson.Document;

import java.util.List;

/**
 * Cross-driver updateUser smoke — Java.
 * <p>
 * Provisions {@code alice_xd} with password {@code orig}, rotates to
 * {@code rotated} via updateUser, then asserts the old password no
 * longer authenticates and the new one does.
 */
public final class UpdateUserSmoke {

    private UpdateUserSmoke() {}

    public static void main(String[] args) {
        String uri = System.getenv("MONGODB_URI");
        String adminPwd = System.getenv("ADMIN_PASSWORD");
        if (uri == null || adminPwd == null) {
            System.err.println("MONGODB_URI and ADMIN_PASSWORD required");
            System.exit(2);
        }

        try (MongoClient root = clientFor(uri, "root", adminPwd, "admin")) {
            root.getDatabase("admin").runCommand(new Document()
                .append("createUser", "alice_xd")
                .append("pwd", "orig")
                .append("roles", List.of(new Document()
                    .append("role", "read")
                    .append("db", "admin"))));
            root.getDatabase("admin").runCommand(new Document()
                .append("updateUser", "alice_xd")
                .append("pwd", "rotated"));
        }

        // Old password — must fail.
        boolean oldFailed = false;
        try (MongoClient old = clientFor(uri, "alice_xd", "orig", "admin")) {
            old.getDatabase("admin").runCommand(new Document("ping", 1));
        } catch (MongoSecurityException | MongoTimeoutException ex) {
            oldFailed = true;
        }
        if (!oldFailed) {
            System.err.println("old password should not authenticate");
            System.exit(1);
        }

        // New password — must succeed.
        try (MongoClient fresh = clientFor(uri, "alice_xd", "rotated", "admin")) {
            fresh.getDatabase("admin").runCommand(new Document("ping", 1));
        }

        System.out.println("OK");
    }

    private static MongoClient clientFor(String uri, String user, String pwd, String authSource) {
        MongoCredential credential = MongoCredential.createScramSha256Credential(
            user, authSource, pwd.toCharArray());
        MongoClientSettings settings = MongoClientSettings.builder()
            .applyConnectionString(new ConnectionString(uri))
            .credential(credential)
            .applyToClusterSettings(b -> b.serverSelectionTimeout(30,
                java.util.concurrent.TimeUnit.SECONDS))
            .build();
        return MongoClients.create(settings);
    }
}
